src/sdi-refresh-monitor.c

```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
+    g_autoptr(SdiSnap) snap = find_snap(self, *p);
+    // Only show progress bar if that snap shown an 'inhibited' notification
+    if (snap == NULL)
+      continue;
+
+    if (!sdi_snap_get_inhibited(snap))
+      continue;
+
+    if (done || hold) {
+      remove_snap(self, snap);
+      SnapRefreshData *data = g_malloc0(sizeof(SnapRefreshData));
+      data->snap_name = g_strdup(*p);
+      data->self = g_object_ref(self);
+      if (done) {
+        snapd_client_get_snap_async(self->client, *p, NULL, show_snap_completed,
+                                    data);
+      }
+      continue;
+    }
+
+    if (sdi_snap_get_hidden(snap) || sdi_snap_get_manually_hidden(snap)) {
+      continue;
+    }
+
+    g_autoptr(SdiRefreshDialog) dialog = sdi_snap_get_dialog(snap);
+    if (dialog == NULL) {
+      // if there's no dialog, get the data for this snap and create
+      // it avoid refresh notifications while the progress dialog is shown
+      sdi_snap_set_ignored(snap, TRUE);
+      SnapRefreshData *data = g_malloc0(sizeof(SnapRefreshData));
+      data->snap_name = g_strdup(*p);
+      data->self = g_object_ref(self);
+      snapd_client_get_snap_async(self->client, *p, NULL,
+                                  begin_application_refresh, data);
+      continue;
+    }
+
+    // if there's already a dialog for this snap, just refresh the
+    // progress bar
+    GPtrArray *tasks = snapd_change_get_tasks(change);
+    gint done = 0;
+    g_autoptr(SnapdTask) current_task = NULL;
+    for (guint i = 0; i < tasks->len; i++) {
+      SnapdTask *task = (SnapdTask *)tasks->pdata[i];
+      const gchar *status = snapd_task_get_status(task);
+      if (status_is_done(status)) {
+        done++;
+      } else if ((current_task == NULL) && g_str_equal("Doing", status)) {
+        current_task = g_object_ref(task);
+      }
+    }
+    if (current_task != NULL) {
+      sdi_refresh_dialog_set_n_tasks_progress(
+          dialog, snapd_task_get_summary(current_task), done, tasks->len);
+    }
+  }
+}
+
+static void update_dock_bar(gpointer key, gpointer value, gpointer data) {
+  SnapProgressTaskData *task_data = value;
+  g_autoptr(SdiRefreshMonitor) self = data;
+
+  gdouble progress =
+      ((gdouble)task_data->done_tasks) / ((gdouble)task_data->total_tasks);
+  if ((progress == task_data->old_progress) && !task_data->done) {
+    return;
+  }
+  task_data->done_tasks = 0;
+  task_data->total_tasks = 0;
+  task_data->old_progress = progress;
+  g_autoptr(GVariantBuilder) builder =
+      g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+  g_variant_builder_add(builder, "{sv}", "progress",
+                        g_variant_new_double(progress));
+  g_variant_builder_add(builder, "{sv}", "progress-visible",
+                        g_variant_new_boolean(!task_data->done));
+  // values has a floating reference, so it's consumed when calling
+  // EMIT_UPDATE()
+  GVariant *values = g_variant_builder_end(builder);
```

Comment:
But what if `task_data->desktop_files->len == 0` won't it leak then? Better to take the reference and ensure it's always cleaned up.

Reply:
Good point. Done.